hoodie-common/src/main/java/com/uber/hoodie/common/util/SerializationUtils.java

```diff
@@ -16,156 +16,126 @@
     try {
-      // stream closed in the finally
-      in = new ObjectInputStream(inputStream);
-      @SuppressWarnings("unchecked") // may fail with CCE if serialised form is incorrect
-      final T obj = (T) in.readObject();
-      return obj;
-
-    } catch (final ClassCastException ex) {
-      throw new HoodieSerializationException("cannot cast class", ex);
-    } catch (final ClassNotFoundException ex) {
-      throw new HoodieSerializationException("class not found", ex);
-    } catch (final IOException ex) {
-      throw new HoodieSerializationException("unable to deserialize to object", ex);
+      return (T) serializerInstance.deserialize(objectData);
     } finally {
-      try {
-        if (in != null) {
-          in.close();
-        }
-      } catch (final IOException ex) { // NOPMD
-        // ignore close exception
-      }
+      releaseSerializerInstance(serializerInstance);
     }
   }
 
-  /**
-   * <p>
-   * Deserializes a single {@code Object} from an array of bytes.
-   * </p>
-   *
-   * <p>
-   * If the call site incorrectly types the return value, a {@link ClassCastException} is thrown from the call site.
-   * Without Generics in this declaration, the call site must type cast and can cause the same ClassCastException.
-   * Note that in both cases, the ClassCastException is in the call site, not in this method.
-   * </p>
-   *
-   * @param <T> the object type to be deserialized
-   * @param objectData the serialized object, must not be null
-   * @return the deserialized object
-   * @throws IllegalArgumentException if {@code objectData} is {@code null}
-   * @throws HoodieSerializationException (runtime) if the serialization fails
-   */
-  public static <T> T deserialize(final byte[] objectData) {
-    if (objectData == null) {
-      throw new IllegalArgumentException("The byte[] must not be null");
+  // Returns cached kryo serializer instance if available otherwise creates a new copy of it.
+  private static KryoSerializerInstance borrowKryoSerializerInstance() {
+    final KryoSerializerInstance kryoSerializerInstance = SerializationUtils.serializerRef
+        .getAndSet(null);
+    if (kryoSerializerInstance == null) {
+      return new KryoSerializerInstance();
+    }
+    return kryoSerializerInstance;
+  }
+
+  // Adds kryo serializer into local cache
+  private static void releaseSerializerInstance(KryoSerializerInstance kryoSerializerInstance) {
+    SerializationUtils.serializerRef.compareAndSet(null, kryoSerializerInstance);
+  }
+
+  private static class KryoSerializerInstance implements Serializable {
+
+    private final Kryo kryo;
+    // Caching ByteArrayOutputStream to avoid recreating it for every operation
+    private final ByteArrayOutputStream baos;
+
+    KryoSerializerInstance() {
+      EmptyScalaKryoInstantiator kryoInstantiator = new EmptyScalaKryoInstantiator();
+      kryo = kryoInstantiator.newKryo();
+      baos = new ByteArrayOutputStream(1024);
```

Comment:
pull 1024 into a constant and consider if this needs to be larger by default.. Looks like the simplest payload will be bigger?

Reply:
Sg. Will change it to 1M. 1M should be sufficient for most cases.